File: Editorial/0091_Decode_Ways/TypeScript/3_Iterative_Constant_Space.ts
function numDecodings(s: string): number {
    if (s.charAt(0) === "0") {
        return 0;
    }
    let n = s.length;
    let twoBack = 1;
    let oneBack = 1;
    for (let i = 1; i < n; i++) {
        let current = 0;
        if (s.charAt(i) !== "0") {
            current = oneBack;
        }
        let twoDigit = parseInt(s.slice(i - 1, i + 1));
        if (twoDigit >= 10 && twoDigit <= 26) {
            current += twoBack;
        }
        twoBack = oneBack;
        oneBack = current;
    }
    return oneBack;
}

File: Editorial/0091_Decode_Ways/CSharp/3_Iterative_Constant_Space.cs
public class Solution {
    public int NumDecodings(string s) {
        if (s[0] == '0') {
            return 0;
        }

        int n = s.Length;
        int twoBack = 1;
        int oneBack = 1;
        for (int i = 1; i < n; i++) {
            int current = 0;
            if (s[i] != '0') {
                current = oneBack;
            }

            int twoDigit = int.Parse(s.Substring(i - 1, 2));
            if (twoDigit >= 10 && twoDigit <= 26) {
                current += twoBack;
            }

            twoBack = oneBack;
            oneBack = current;
        }

        return oneBack;
    }
}
